### PR TITLE
PBP dataset performance enhancements

### PR DESCRIPTION
**Pull request by M-R-Schaefer**
I have reworked the PBP dataset to enhance performance.
Previously, the buffer would be filled with batches using multiprocessing, run dry and then be refilled.
Now, there is a dedicated processing thread which uses MP internally.
That means the buffer is now continuously filled without blocking train step compute.

On my home PC I have achieved a 20 % epoch time reduction compared to the cached dataset.
On my work laptop, it's roughly 30 % slower, but still much faster than the old pbp (old: 6.3 s, new: 1.3 s)

Apart from this dataset, there is also the CSVLogger class which uses tensorflow. 
I think if we write our own logger, we could turn tensorflow into an optional dependency.

TODO:

- [x] add vesin to poetry
- [x] swap out matscipy for vesin everywhere
- [x] remove matscipy dependency
- [x] make sure tests are running
- [x] rerun benchmark
- [x] add stress to batch processor

Update:

After rerunning the benchmark on my workstation, the cached dataset is still faster than the PBB, especially in worst case scenarios (same system size, low batch size, lightweight model).
300 atom system, batch size 4, rmax=6

cached: 1.5 s/epoch
pbp new : 2.5 s/epoch
pbp old: 4.7 s/epoch

With BS 8, n_ens = 8, rmax=5
cached: 5.6 s/epoch
pbp: 6.7 s/epoch

### CODE REVIEW

File: apax/data/input_pipeline.py [hunk 1]
```diff
@@ -358,88 +360,82 @@ def next_power_of_two(x):
...
+        offsets = []
+        for i, (inp, lab) in enumerate(samples):
+            inputs["numbers"][i, : inp["n_atoms"]] = inp["numbers"]
+            inputs["n_atoms"][i] = inp["n_atoms"]
+            inputs["positions"][i, : inp["n_atoms"]] = inp["positions"]
+            inputs["box"][i] = inp["box"]
 
-        max_atoms = next_power_of_two(max_atoms)
-        max_nbrs = next_power_of_two(max_nbrs)
+            idx, offset = compute_nl(inp["positions"], inp["box"], self.cutoff)
```
Comment (PythonFZ) [inline apax/data/input_pipeline.py:415]: You do not pass the `n_atoms` here so `compute_nl` computes the full neighbourlist, even if some positions are `zero` from padding? Wouldn't it be saver to not zero-pad but use NaN and filter? Is that slower? Can it be an issue if an atom is exactly at `(0, 0, 0)` which can happen quite frequently?

File: apax/data/input_pipeline.py [hunk 2]
```diff
@@ -358,88 +360,82 @@ def next_power_of_two(x):
...
+            idxs.append(idx)
+            offsets.append(offset)
 
-        for i in range(len(inputs["n_atoms"])):
-            inputs["idx"][i], inputs["offsets"][i] = pad_nl(
-                inputs["idx"][i], inputs["offsets"][i], max_nbrs
-            )
+            labels["energy"][i] = lab["energy"]
+            if self.forces:
+                labels["forces"][i, : inp["n_atoms"]] = lab["forces"]
```
Comment (PythonFZ) [inline apax/data/input_pipeline.py:421]: How are other labels treated, e.g. `stress`?

File: apax/data/input_pipeline.py [hunk 3]
```diff
@@ -358,88 +360,82 @@ def next_power_of_two(x):
...
 
-            zeros_to_add = max_atoms - inputs["numbers"][i].shape[0]
-            inputs["positions"][i] = np.pad(
-                inputs["positions"][i], ((0, zeros_to_add), (0, 0)), "constant"
-            )
-            inputs["numbers"][i] = np.pad(
-                inputs["numbers"][i], (0, zeros_to_add), "constant"
-            ).astype(np.int16)
+        max_nbrs = np.max([idx.shape[1] for idx in idxs])
+        max_nbrs = next_power_of_two(max_nbrs)  # TODO better intervals
```
Comment (PythonFZ) [inline apax/data/input_pipeline.py:424]: TODO: better intervals?

File: apax/data/input_pipeline.py [hunk 4]
```diff
@@ -358,88 +360,82 @@ def next_power_of_two(x):
...
+        max_nbrs = round_up_to_multiple(max_nbrs, self.nl_padding)
 
-            if "forces" in labels:
-                labels["forces"][i] = np.pad(
-                    labels["forces"][i], ((0, zeros_to_add), (0, 0)), "constant"
-                )
+        inputs["idx"] = np.zeros((n_samples, 2, max_nbrs), dtype=np.int16)
```
Comment (PythonFZ) [inline apax/data/input_pipeline.py:427]: int16 goes to `32767` does that mean structures with more than that can not be trained on? That is not relevant for DFT but if we ever consider coarse grained simulations we might want to keep that in mind? But I am also not entirely shure what `idx` from the nl actually is.

File: apax/data/input_pipeline.py [hunk 5]
```diff
@@ -488,61 +478,86 @@ def __init__(
...
+        self.prepare_batch = BatchProcessor(
+            cutoff, forces, stress, atom_padding, nl_padding
+        )
 
         self.count = 0
-        self.reset_every = reset_every
+
         self.max_count = self.n_epochs * self.steps_per_epoch()
-        self.buffer = deque()
+        from queue import Queue
```
Comment (PythonFZ) [inline apax/data/input_pipeline.py:488]: Do we want/need the import here or can it be moved to the module level?

File: apax/md/ase_calc.py [hunk 6]
```diff
@@ -9,8 +9,8 @@
 from ase.calculators.calculator import Calculator, all_changes
 from ase.calculators.singlepoint import SinglePointCalculator
 from flax.core.frozen_dict import freeze, unfreeze
-from matscipy.neighbours import neighbour_list
 from tqdm import trange
+from vesin import NeighborList
```
Comment (PythonFZ) [inline apax/md/ase_calc.py:13]: `vesin` > `matscipy`. Do we still need `matscipy` elsewhere? I think it can be removed then.

**Review by PythonFZ — COMMENTED** [review]
Some short questions. I am struggling a bit giving a good review because I am not very familiar with the multiprocessing in such extend.

File: apax/data/input_pipeline.py (hunk 1 above)
Comment (M-R-Schaefer) [reply apax/data/input_pipeline.py:415]: `inp["positions"]` are the unpadded positions. Padded positions are masked during model calculations not by their 0,0,0 value, but by the atomic number 0 in the `numbers` array.

File: apax/data/input_pipeline.py (hunk 2 above)
Comment (M-R-Schaefer) [reply apax/data/input_pipeline.py:421]: oh, I removed the inclusion of stress... Good catch

File: apax/data/input_pipeline.py (hunk 3 above)
Comment (M-R-Schaefer) [reply apax/data/input_pipeline.py:424]: line 424 needs to be deleted. `max_nbrs` gets overwritten in the next line anyway, which is the IMO more sensible approach.
Powers of two are pretty wasteful. The input now allows the next integer (default 2000) to pad neighbors to.

File: apax/data/input_pipeline.py (hunk 4 above)
Comment (M-R-Schaefer) [reply apax/data/input_pipeline.py:427]: That is true. I switched it out to int16 at some point because it cut down memory requirements by like 15 % or so. This was during the last update to the input pipeline when we tried to figure out how to train on large datasets.
I am not sure it will be relevant for training on coarse grained structures, as 33 K atoms might be too large of a structure for a GPU to handle.

idx is the sparse [2,n_neighbors] neighborlist.

File: apax/data/input_pipeline.py (hunk 5 above)
Comment (M-R-Schaefer) [reply apax/data/input_pipeline.py:488]: nope, I'll move it up

File: apax/md/ase_calc.py (hunk 6 above)
Comment (PythonFZ) [reply apax/md/ase_calc.py:13]: I think I reviewed before you made the changes to pyproject.toml

File: apax/data/input_pipeline.py (hunk 3 above)
Comment (M-R-Schaefer) [reply apax/data/input_pipeline.py:424]: oh, that means I need to rerun the benchmark.... Currently, the padding was applied twice. The NL was first padded to the next power of 2 and then to the next integer.

Comment (M-R-Schaefer) [reply apax/data/input_pipeline.py:424]: so maybe, the pbp dataset will close the performance gap with the cached one afterall.

**Review by PythonFZ — APPROVED** [review]
LGTM 👍
